packages/react-router/index.js

```diff
@@ -97,7 +97,9 @@ if (__DEV__) {
  */
 export function Navigate({ to, replace = false, state }) {
   let navigate = useNavigate();
-  navigate(to, { replace, state });
+  React.useEffect(() => {
+    navigate(to, { replace, state });
+  }, [navigate, replace, state, to]);
```

Comment:
this makes sense

Reply:
It breaks synchronous redirects on the initial render.

Reply:
It shouldn't be synchronous, though. If there is a Redirect rendered, that navigation should happen in a second render pass. That also lets CM break up that work between frames (or not).

Reply:
What about `useLayoutEffect`? After all that's what it's able to do: synchronous side effects.

Reply:
useLayoutEffect happens *after* DOM manipulations, so effectively the render has already occurred. Also, those will produce errors in SSR, so you have to build detection to swap out for useEffect (which is a whole can of worms...).

Reply:
> useLayoutEffect happens after DOM manipulations

I know that but it's still synchronous so that there won't be two different paints. I thought that was the original goal: to not paint a screen you'll redirect from.

Reply:
Sorry for the delay. After thinking about this a lot I think we just can't support redirect on the initial render once we get into the React lifecycle. In order to do this properly, we'd have to throw.

I'm going to change the tests around this and update our recommendation in v6 so people know if they need to redirect on the initial render, they should probably be doing it on the server.